tensornetwork/block_sparse/charge.py

```diff
@@ -114,3 +114,272 @@ def __repr__(self):
+    """
+    Fuse `self` with `other`.
+    Args:
+      other: A `BaseCharge` object.
+    Returns:
+      BaseCharge: The result of fusing `self` with `other`.
+    """
+
+    # fuse the unique charges from each index, then compute new unique charges
+    comb_charges = fuse_ndarray_charges(self.unique_charges,
+                                        other.unique_charges, self.charge_types)
+    [unique_charges, charge_labels] = np.unique(
+        comb_charges, return_inverse=True, axis=1)
+    charge_labels = charge_labels.reshape(self.unique_charges.shape[1],
+                                          other.unique_charges.shape[1]).astype(
+                                              np.int16)
+
+    # find new labels using broadcasting
+    charge_labels = charge_labels[(
+        self.charge_labels[:, None] +
+        np.zeros([1, len(other)], dtype=np.int16)).ravel(), (
+            other.charge_labels[None, :] +
+            np.zeros([len(self), 1], dtype=np.int16)).ravel()]
+
+    obj = self.__new__(type(self))
+    obj.__init__(unique_charges, charge_labels, self.charge_types)
+
+    return obj
+
+  def dual(self, take_dual: Optional[bool] = False) -> "BaseCharge":
+    """
+    Return the charges of `BaseCharge`, possibly conjugated.
+    Args:
+      take_dual: If `True` return the dual charges. If `False` return 
+        regular charges.
+    Returns:
+      BaseCharge
+    """
+    if take_dual:
+      unique_dual_charges = np.stack([
+          self.charge_types[n].dual_charges(self.unique_charges[n, :])
+          for n in range(len(self.charge_types))
+      ],
+                                     axis=0)
+
+      obj = self.__new__(type(self))
+      obj.__init__(unique_dual_charges, self.charge_labels, self.charge_types)
+      return obj
+    return self
+
+  def __matmul__(self, other):
+    #some checks
+    if len(self) != len(other):
+      raise ValueError(
+          '__matmul__ requires charges to have the same number of elements')
+    charges = np.concatenate([self.charges, other.charges], axis=0)
+    charge_types = self.charge_types + other.charge_types
+    return BaseCharge(
+        charges=charges, charge_labels=None, charge_types=charge_types)
+
+  def __mul__(self, number: bool) -> "BaseCharge":
+    if not isinstance(number, (bool, np.bool_)):
+      raise ValueError(
+          "can only multiply by `True` or `False`, found {}".format(number))
+    return self.dual(number)
+
+
+class U1Charge(BaseCharge):
+
+  def __init__(self,
+               charges: np.ndarray,
+               charge_labels: Optional[np.ndarray] = None,
+               charge_types: Optional[List[Type["BaseCharge"]]] = None) -> None:
+    super().__init__(charges, charge_labels, charge_types=[type(self)])
+
+  @staticmethod
+  def fuse(charge1, charge2) -> np.ndarray:
+    return np.add.outer(charge1, charge2).ravel()
+
+  @staticmethod
+  def dual_charges(charges) -> np.ndarray:
+    return charges * charges.dtype.type(-1)
```

Comment:
What does type(-1) do?

Reply:
type casting (casting -1 to whatever dtype charges have)

Reply:
Aw I see. Ok this should be fine then.